libmamba/include/mamba/core/graph_util.hpp

```diff
@@ -7,38 +7,47 @@
 #ifndef MAMBA_CORE_GRAPH_UTIL_HPP
 #define MAMBA_CORE_GRAPH_UTIL_HPP
 
-#include <map>
 #include <utility>
 #include <vector>
+#include <map>
 
 namespace mamba
 {
     // Simplified implementation of a directed graph
-    // where a path exists between each node and the
-    // first one (you can think of it as a tree with
-    // potential cycles)
-
-    template <class T>
-    class graph
+    template <typename Node, typename GraphCRTP>
+    class DiGraphBase
```

Comment:
If the definition of `DiGraph<Node, void>` remains empty, I don't think we need a CRTP here, simple inheritance should be enough. A CRTP is usually used when you want to provide different implementation of a same interface without relying on virtual method and entity semantics, and / or when you want to avoid a catch-all template generic function, and / or you don't want to lose the final type information and / or a combination of the three.

Reply:
Actually the CRTP is justified here, since a downcast is required in the `depth_first_search` method before invoking the visitor (I missed it during the first review). A simple inheritance would make the downcast more fragile since it would require to use the visitor internal `graph_t` type

Reply:
@JohanMabille We could kill the CRTP if we change the visitor to be responsible for its own capture of the graph. Would be even easier with a lambda. Note there is currently only one DFS visitor implemented in mamba right now, but there will be more with the error message algo.

Thinking about this as I am about to add other visitor patterns (*e.g.* for leaves) that don't need to have so many options.
(Ideally we'd have a range but let's keep it simple.)

Reply:
> We could kill the CRTP if we change the visitor to be responsible for its own capture of the graph.

Not sure to get what you mean. The base graph class would not invoke the visitor anymore? The only way I see to do this would be to have public tempate functions in the final classes that forward to a common private implementation with the appropriate class, but that's potentially a lot of boilerplate.

Reply:
Now we have:
```cpp
auto graph = DiGraph();
struct Visitor {
   void start_node(node_id n, graph_t const& g);
};
graph.depth_first_search(Visitor());
```
We could have (which is kind of ugly but gets rid of CRTP)
```cpp
auto graph = DiGraph();
struct Visitor {
   graph_t& g;

   void start_node(node_id n);
};
graph.depth_first_search(Visitor(g));
```
Or better with a lambda (swapping the different callback names for a `where` enum):
```cpp
auto graph = DiGraph();
graph.depth_first_search([&graph](auto where, node_id n) {
  if (what == Search::start_node){
    ...
  }
});
```
And if we really care about having separate functions in the visitor, we could have the visitor's `operator()` be templated by the `Search` enum, but that brings back the verbosity (at least until C++20 templated lambda).

Reply:
I find it more verbose and dangerous, you need to pas to the visitor the same instance of graph as that used for calling the method that will invoke the visitor. Besides, it is less flexible since you cannot reuse the visitor instance for another graph instance (even if I'm not sure we will need this).